web/packages/shared/components/Validation/Validation.tsx

```diff
@@ -0,0 +1,190 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+
+import { Logger } from 'design/logger';
+
+import { isObject } from 'shared/utils/highbar';
+import { Store, useStore } from 'shared/libs/stores';
+
+import { ValidationResult } from './rules';
+
+const logger = new Logger('validation');
+
+/** A per-rule callback that will be executed during validation. */
+type RuleCallback = () => void;
+
+export type Result = ValidationResult | boolean;
+
+type ValidatorState = {
+  /** Indicates whether the last validation was successful. */
+  valid: boolean;
+  /**
+   * Indicates whether the validator has been activated by a call to
+   * `validate`.
+   */
+  validating: boolean;
+};
+
+/** A store that handles input validation and makes its results accessible. */
+export default class Validator extends Store<ValidatorState> {
+  state = {
+    valid: true,
+    validating: false,
+  };
+
+  /**
+   * @deprecated For temporary Enterprise compatibility only. Use {@link state}
+   * instead.
+   */
+  valid = true;
```

Comment:
is this something to just keep the builds from breaking? I'd add a TODO here to ensure it gets updated after, otherwise this could get messy. I wonder if it'd be better to use a `getter` here to always return the value of `state` instead of a separate property. like this pseudo-ish code
```js
  get valid() {
    return this.state.valid;
  }
  ```

Reply:
The cleanup PR https://github.com/gravitational/teleport/pull/49550 removes it, but I'll follow your advice anyway.

Reply:
Ah you're right. I think we can ignore this comment then

Reply:
Actually I wanted to suggest using a getter as well. Accessing `state.valid` in the callsites look a bit off, like accessing an internal property of `Validator`. I believe having getters for `valid` and `validating` would provide a nicer API.

But I will leave it up to you.